backend/daemon/api/documents/v1alpha/documents.go

```diff
@@ -947,6 +947,85 @@ func (api *Server) ListAccountPublications(ctx context.Context, in *documents.Li
 	return out, nil
 }
 
+// MergeChanges implements the corresponding gRPC method. It merges changes and publishes them.
+func (api *Server) MergeChanges(ctx context.Context, in *documents.MergeChangesRequest) (*documents.Document, error) {
+	me, err := api.getMe()
+	if err != nil {
+		return nil, err
+	}
+
+	if in.TargetDocumentId == "" {
+		return nil, fmt.Errorf("Target ID is mandatory")
+	}
+	eid := hyper.EntityID(in.TargetDocumentId)
+
+	source, err := hyper.Version(in.SourceDocumentVersion).Parse()
+	if err != nil {
+		return nil, err
+	}
+
+	doc, err := api.CreateDraft(ctx, &documents.CreateDraftRequest{
+		ExistingDocumentId: in.TargetDocumentId,
+		Version:            in.TargetDocumentVersion,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if doc.Id != eid.String() {
+		return nil, fmt.Errorf("Internal error. Draft entity and doc don't match")
+	}
+	entity, err := api.blobs.LoadEntity(ctx, eid)
+	if err != nil {
+		return nil, err
+	}
+	ch, err := api.blobs.GetDraft(ctx, eid)
+	if err != nil {
+		return nil, err
+	}
+
+OUTER:
+	for _, dep := range source {
+		for _, alreadyDep := range ch.Deps {
+			if dep == alreadyDep {
+				continue OUTER
+			}
+		}
+		ch.Deps = append(ch.Deps, dep)
+	}
+
+	del, err := api.getDelegation(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	hb, err := entity.CreateChange(entity.NextTimestamp(), me.DeviceKey(), del, map[string]any{
```

Comment:
The rule of 

> _If there's only one dep, and no content is being changed, then we need to abort this change. But if there're more than one dep, it means that we are merging, and we should let this change pass through_ 

does not work since when I call `CreateChange` in order to see the number of deps is by inspecting the heads of the entity even though I create that change with `hyper.WithDeps(ch.Deps)` including those changes.
Maybe the way I include those changes is not the correct one...

Reply:
> when I call CreateChange in order to see the number of deps is by inspecting the heads of the entity even though I create that change with hyper.WithDeps(ch.Deps) including those changes.

I don't understand this. What exactly doesn't work, and what errors do you get?

I think the following should work (basically do the same that GetPublication would do with a combined version):


1. Receive a list of version strings to be merged.
2. Create an empty set to take the union of all the CIDs from all the versions to be merged.
3. Parse each version (we have a function for that) and collect all the CIDs into the set.
5. Load the entity using the CIDs from the set as heads.
6. From the loaded entity instance create a new change (no need to specify deps explicitly, the entity will figure out the deps automatically, because some of the versions you've collected may be redundant).

Reply:
I thought that at the beginning but after calling `LoadEntityFromHeads` with all those CID's I have a nil entity. No errors but nil entity because the entity in step 4 does not have any changes `len(changes) == 0`. @burdiyan

Reply:
Then there's nothing to merge? I don't get it. If you try to call `LoadEntityFromHeads` it won't load anything. These entity loading functions would return `nil, nil` in case of "not found". I regretted this design a few times, but didn't bother changing it, as in many places we explicitly check for this case.